### PR TITLE
[WIP] Refactor graph models

### PR DESCRIPTION
**Pull request by peastman**
This fixes the problems that were discussed in #1218.  So far I've only changed GraphConvModel.  If the changes look good to everyone, I'll make similar changes to the other graph models.  Here are the effects of these changes:

- Uncertainty prediction is now implemented.
- I added a `n_classes` option for classification models.  Before it was hardcoded to only allow two classes.
- Training should be faster for multitask models with lots of tasks.
- All prediction methods produce consistent (and hopefully now correct!) outputs.
- Dropout should work better than before.

There's some API inconsistency between the graph models and fully connected models.  MultiTaskClassifier and MultiTaskRegressor have an argument called `dropouts` (plural) which can be either a float (use the same dropout rate for all layers) or a list (specify the rate separately for all layers).  It defaults to 0.5.  In contrast, GraphConvModel and PetroskiSuchModel have an argument called `dropout` (singular) which can only be a float.  It defaults to 0.0.  And then WeaveModel, DTNNModel, DAGModel, and MPNNModel don't support dropout at all.  Should I change these to be more consistent?  In particular,

- Should I add dropout to the models that don't currently support it?
- Should I allow you to specify a list rather than a float?
- Should I change the argument name to `dropouts`?
- Should I change the default value to 0.5?

### CODE REVIEW

**Comment by lilleswing** [comment]
* Should I add dropout to the models that don't currently support it?
Yeah I think so.
* Should I allow you to specify a list rather than a float?
We should follow the standard you set in MultiTaskModel and allow it to take either.
* Should I change the argument name to dropouts?
Yes lets take the hit now and get consistency for later
* Should I change the default value to 0.5?
I am torn on this.  I like not having dropout by default but it means we can't have uncertainty predictions by default.  Given the choice between the two I think it is more important to be able to have uncertainty predictions so yes default to 0.5.

**Comment by peastman** [comment]
> but it means we can't have uncertainty predictions by default.

You don't get uncertainty by default anyway: you have to specify `uncertainty=True`.  And if you do that without enabling dropout, it throws an exception telling you to enable dropout.

**Comment by lilleswing** [comment]
Ah yes you are right -- if uncertainty is already an additional flag I prefer defaulting to no dropout in models.

**Comment by rbharath** [comment]
@peastman Like the design! I think @lilleswing hit all the important points, so good on my end.

**Comment by peastman** [comment]
I'm totally confused by DAGModel.  Instead of being a stack of layers in the normal way, it contains just a single DAGLayer, which internally implements an effective stack of layers, all contained within one "layer".  That "layer" has an argument `layer_sizes` you can use to specify how many "internal layers" it should have, along with their widths.  But DAGModel doesn't provide any means to set that property.  It always just uses the default (a single layer of size 100, although the docstring incorrectly claims it's a single layer of size 1000).  It also has an option to specify a dropout probability, but it's ignored.

What is this class supposed to do?  What is it used for?

Which of the graph models are actually functional and supported?

**Comment by rbharath** [comment]
@peastman I think all the models on MoleculeNet.ai should be functional and supported, although I suspect that some are more robust than others. 

@miaecle should be able to answer questions about the `DAGModel`. I recall that this model was very complicated and painful to implement in TensorFlow. These complexities might be why the design isn't very modular.

@lilleswing Should be able to comment on `PetroskiSuchModel`. IIRC, this was an experimental model (not in MoleculeNet.ai yet for example)

**Comment by miaecle** [comment]
@peastman Sorry for the weird notation. Basically DAG is more like an RNN that propagate through all directed edges. The `layer_sizes` define the NN structure for a single step in the propagation. Each `DAGLayer` will perform one propagation over all edges.(which is the default in its original paper) You could definitely try using multiple `DAGLayer`, though using one layer is already super slow. The `dropout` might inherit from other models and it is totally fine to get rid of it.

**Comment by peastman** [comment]
> @lilleswing Should be able to comment on PetroskiSuchModel. IIRC, this was an experimental model (not in MoleculeNet.ai yet for example)

If it's not fully functional or supported, should we move it to contrib?

Should DAGModel expose the `layer_sizes` option?  What should the default value be?  `[100]`, as it currently is, or `[1000]`, as the docstring claims it is?  At what point should dropout be applied?  After each step in `DAGgraph_step()`?  Or just once after the whole DAGLayer?

**Comment by miaecle** [comment]
@peastman Let me make the initial changes for the DAGLayer. I am not sure about the dropout pattern here since it is not a standard process in this model, my opinion would be apply it after each step(propagation through an edge).

**Comment by lilleswing** [comment]
Yes I can move PetroskiSuckModel to contrib

**Comment by peastman** [comment]
I've written test cases for WeaveModel (works ok) and DAGModel (slow but works ok).  But MPNNModel is just crazy slow.  On a toy dataset with only 10 samples, it can take over a minute per epoch.  I based the code on delaney_MPNN.py.  Is it supposed to be like that?

For DTNNModel, I can't even figure out what the inputs are supposed to be.  Can anyone provide insight?

**Comment by rbharath** [comment]
As a drive-by comment, I suspect this is a normalization issue...

**Comment by peastman** [comment]
No, it actually was just training with far too low a learning rate for far too few epochs.  If I multiply the learning rate by 1000 and the number of epochs by 10, it gives a much better fit.

This leaves the question, why is the correlation coefficient so good even when the model has barely been trained at all and the output is totally wrong?  @miaecle do you have any idea about this?

**Comment by peastman** [comment]
This should be ready to merge once the tests pass.  All the graph models have been refactored as described in #1214.  DAGModel, GraphConvModel, and MPNNModel now support dropout and uncertainty prediction.  DTNN model supports dropout, but I haven't been able to get uncertainty to work.  And I couldn't get even dropout to work with WeaveModel.  Adding even a tiny amount of dropout to it seems to totally break the model.  These might just be artifacts of the artificial test cases I was using, but I don't want to add features until I can create test cases that show they work.

**Comment by peastman** [comment]
Ok, this is ready to merge.
